receiver/cudareceiver/collector.go

```diff
@@ -0,0 +1,191 @@
+// Copyright 2020, OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cudareceiver
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"contrib.go.opencensus.io/resource/auto"
+	metricspb "github.com/census-instrumentation/opencensus-proto/gen-go/metrics/v1"
+	resourcepb "github.com/census-instrumentation/opencensus-proto/gen-go/resource/v1"
+	"github.com/golang/protobuf/ptypes"
+	"go.opentelemetry.io/collector/consumer"
+	"go.opentelemetry.io/collector/consumer/consumerdata"
+	"go.uber.org/zap"
+)
+
+var (
+	rsc                   *resourcepb.Resource
+	resourceDetectionSync sync.Once
+)
+
+// CUDAMetricsCollector is a struct that collects and reports CUDA (GPU) metrics (temprature, power, et al.).
+//
+// TODO: ConsumeMetricsOld will be deprecated and should be replaced with ConsumerMetrics.
+type CUDAMetricsCollector struct {
+	consumer consumer.MetricsConsumerOld
+
+	startTime time.Time
+	device    *NVMLDevice
+
+	scrapeInterval time.Duration
+	metricPrefix   string
+	done           chan struct{}
+
+	logger *zap.Logger
+}
+
+// NewCUDAMetricsCollector creates a new set of CUDA (GPU) Metrics (temperature, power, et al.).
+func NewCUDAMetricsCollector(d time.Duration, prefix string, logger *zap.Logger, con consumer.MetricsConsumerOld) (*CUDAMetricsCollector, error) {
+	device, status := NVMLDeviceGetHandledByIndex(uint64(0))
+	if status != NVMLSuccess {
+		return nil, fmt.Errorf("could not get GPU device: status=%d", status)
+	}
+
+	c := &CUDAMetricsCollector{
+		consumer:       con,
+		startTime:      time.Now(),
+		device:         device,
+		scrapeInterval: d,
+		metricPrefix:   prefix,
+		done:           make(chan struct{}),
+		logger:         logger,
+	}
+
+	return c, nil
+}
+
+func detectResource() {
```

Comment:
You can probably drop this resource detection piece that depends on an OC package and just rely on the [resource detection processor](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/processor/resourcedetectionprocessor) instead, although note the resource detection processor needs to be extended to include more types of environments

(see related discussion here if interested: https://github.com/open-telemetry/opentelemetry-collector/issues/871)

Reply:
It's great to have the resource detector already in SDK. Can I expect all resource information is included in `OTEL_RESOURCE` env variable as the [README of `resourcedetectionprocessor`](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/master/processor/resourcedetectionprocessor) says, and just use `os.Getenv()` instead of `auto.Detect()` here?

Reply:
The resource detection processor will read resource info out of the `OTEL_RESOURCE` env variable (or from GCE metadata, etc) and put it into the metric data.

If you want to use it you don't need to do anything here. Just need to setup a pipeline like this:
```yaml
receivers:
  cudareceiver:
    ...

processors:
  resourcedetection:
    # runs everything by default iirc

service:
  pipelines:
    metrics:
      receivers: [cudareceiver]
      processors: [resourcedetection]
      exporters: [...]
```

Reply:
Oh, I took it incorrectly. So, that means, in the receiver we don't need to put anything into resource that is used `metricspb.Metric` unless we put extra labels into resources, correct?